### ISSUE TITLE
Exclude HTML5 data attributes from being removed

### ISSUE THREAD

**Issue by manish-in-java** [OWASP/java-html-sanitizer#68]
This is the same as issue #51, since that one has been closed without a resolution. I would like to allow users to provide HTML5 code such as:

    <a data-target="..." data-action="..." data-profile="...">...</a>

Basically, any number of attributes on an HTML5 tag (`a` in this example) that begin with `data-`. Currently, passing such HTML through the Java sanitizer strips the `data-` attributes. The attribute names are not known in advance and there can be a huge number of attributes like this so hard-coding allowed attributes in the policy is not possible.

How do we preserve these attributes with `HtmlPolicyBuilder` (or otherwise)? Tried `.allowAttributes("data-*").onElements("a")` but that did not work. The `allowAttributes` method performs an exact match on the attribute name, which obviously does not work because the attribute name is intended to be a pattern instead of an exact name.

Please provide exact examples or code and do not close this issue with a generic answer like that provided for issue #51.

**Comment by mikesamuel** [comment]
How does one know whether a data attribute is a URL?

It seems allowing a data-action with a `javascript:` URL would be bad.

**Comment by jmanico** [comment]
While we do not support wildcards, but you CAN permit multiple attributes.

.allowAttributes("data-target", "data-action").onElements("a")

**Comment by jmanico** [comment]
Since this feature is already supported in the current policy, we're not likely to augment with wildcards. Closing for now, but re-open if you need something that the current policy builder does not support.

**Comment by Mirodil** [comment]
@manish-in-java how did you solve this issue?

**Comment by manish-in-java** [comment]
@Mirodil, I have stopped using the OWASP Sanitizer.

**Comment by Mirodil** [comment]
I see, Thank you

**Comment by osamesama** [comment]
Bumping into the same problem.  I am building an web service that, among other things, allows CRUD access to certain objects that contain sanitized HTML snippets.  I desire to let those clients use `data-` attributes for their own purposes, and I don't want to restrict them to an explicit list that I determine, since their needs vary so much.  I want my clients to use any `data-` attributes that fit their needs.  This seems to be a challenge with this library.

**Comment by mikesamuel** [comment]
@osamesama Same question for you: https://github.com/OWASP/java-html-sanitizer/issues/68#issuecomment-221931410

**Comment by osamesama** [comment]
Answer is, it shouldn't matter.  `data-` attributes are custom fields for the app's own use, and the sanitizer shouldn't assume or promise responsibility for it.  Data attributes can only be exploited if the system uses them in a certain way; they are not inherently risky.  However, any user can misuse a perfectly safe tool and make it unsafe.

At some point, a software system (or car, or hammer, or water) has to give the user a little slack and assume it doesn't always know what's best for the user.  Just because the library can't conceive of or protect from all misuses of `data-` attributes doesn't mean it shouldn't provide an option to permit them if the developer opts-in for it.

After all, it may be a far worse scenario for a developer to abandon the system in frustration and (arrogantly, recklessly) roll his own.

**Comment by mikesamuel** [comment]
>  it shouldn't matter.

Whether it should matter, if JavaScript does something like the below, then it does matter.

```html
<span data-url="javascript:alert(1)">x</span>
<script>
for (const thing of document.querySelectorAll('*[data-url]')) {
  const url = thing.getAttribute('data-url');
  const a = document.createElement('a');
  a.href = url;
  thing.parentNode.insertBefore(a, thing);
  a.appendChild(thing);
}
</script>
```

> they are not inherently risky

Allowing arbitrary `data-`\* attributes without type metadata leaves one open to [script gadgets](https://dl.acm.org/citation.cfm?id=3134091).

"""
HTML Sanitizers: Sanitizers aim at removing potentially malicious content. Most sanitizers do this by defining a known-good list of tags and attributes and removing anything else from a provided string. This list varies from sanitizer to sanitizer. The Closure sanitizer for example, removes data- attributes, while DOMPurify allows them in its default configuration. Furthermore, all sanitizers we looked at allow id and class attributes. Hence, we investigated whether this behavior is secure. In our data set 78.30% of all second level domains had at least one data flow from an HTML attribute into a security-sensitive sink, whereas 59.51% of the sites exhibited such flows from data- attributes. Furthermore, 15.67% executed data from id attributes and 10% from class attributes. **Based on these numbers, we recommend to revisit at least the sanitization approach towards blocking data- attributes.**
"""

This tool aims to provide safe output.  I don't see that distinguishing between *inherently* risky and other types of risky helps towards that goal.

> has to give the <del>user</del><ins>developer</ins> a little slack

If there were one developer in this equation, I might agree.  The point of the script gadgets paper, is that there are specific ways that small mistakes compound to leave many applications vulnerable and that those are systematically exploitable.  Whitelisting custom attributes is one of those mechanisms.

It seems that I would be remiss if I assumed that any developer who can opt-in is aware of the security literature and has coordinated with the other developers who grant privilege to data attributes.

> After all, it may be a far worse scenario for a developer to abandon the system in frustration and (arrogantly, recklessly) roll his own.

I agree, and I hope you don't do that.

But making it easier to unintentionally harm yourself and your users seems a bad scenario.
